### PR TITLE
implement group-norm

### PR DESCRIPTION
**Pull request by GHOpenonic**
implement group and layer normalization with tests. Change norm selection to use build_norm_layer that is more extensible than old if/elif logic.

### CODE REVIEW

File: src/samudra/models/modules/blocks.py [hunk 1]
```diff
@@ -343,6 +338,30 @@ def __init__(
         self.convblock = torch.nn.Sequential(*convblock)
         self.checkpoint_simple = checkpoint_simple
 
+    @staticmethod
+    def _build_norm_layer(
```
Comment (alxmrs) [inline src/samudra/models/modules/blocks.py:342]: It looks like this doesn't use any class members, therefore I think it could be a pure function not in the class rather than a method. Further, I think this could be generally useful -- how about we make it a public function?

File: src/samudra/models/modules/blocks.py [hunk 2]
```diff
@@ -343,6 +338,30 @@ def __init__(
...
+        norm: str, channels: int, group_norm_groups: int
```
Comment (alxmrs) [inline src/samudra/models/modules/blocks.py:343]: I think the parameter could just be `groups`? 

🐑 maybe it could be an optional param since it's only used with group norm.

File: src/samudra/models/modules/blocks.py (hunk 1 above)
Comment (alxmrs) [inline src/samudra/models/modules/blocks.py:342]: 🐑  idea: potentially, we could pass in to the core block a "NormBuilder" so we could set all of the hparams for which norm to use in our config system -- see the `CoreBlockBuilder(Protocol)` at the bottom of this file.

File: src/samudra/config.py [hunk 3]
```diff
@@ -336,7 +336,7 @@ def make_source(
 
 BlockType = Literal["conv_next_block", "conv_block"]
 ActivationType = Literal["relu", "gelu", "capped_gelu"]
-NormType = Literal["batch", "instance", "layer"]
+NormType = Literal["batch", "instance", "group", "nonorm", "layer"]
```
[the remark this answers is not in the document]
Comment (alxmrs) [reply src/samudra/config.py:339]: See my 🐑 suggestion above about the NormBuilder -- that idea could help address this concern.

File: tests/test_blocks.py [hunk 4]
```diff
@@ -18,7 +19,7 @@
     height=st.integers(min_value=1, max_value=16),
     width=st.integers(min_value=1, max_value=16),
 )
-@settings(max_examples=20)
+@settings(max_examples=20, deadline=None)
```
Comment (alxmrs) [inline tests/test_blocks.py:22]: I think we should remove this -- if there's a timeout in our property based tests, that's a concern worth knowing about.

**Review by alxmrs — COMMENTED** [review]
A few ideas that could improve this group norm PR.

**Comment by jder** [comment]
I have included this in #671 which I'm planning to merge soon.
